src/chip/mod.rs

```diff
@@ -0,0 +1,45 @@
+//! Quilc-compatible chip specification
+use serde::Serializer;
+use serde_aux::prelude::deserialize_string_from_number;
+
+pub mod isa;
+mod parity_test;
+pub mod specs;
+
+/// A `ChipSpec` defines the various hardware objects that are available
+/// when compiling a quantum program.
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub struct ChipSpec {
+    /// The "Instruction Set Architecture" of the chip; i.e. the
+    /// instructions (or "gates") that a particular hardware object
+    /// supports.
+    pub isa: isa::Isa,
+    /// The various operating characteristics of a hardware object (e.g
+    /// readout fidelity).
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub specs: Option<specs::Specs>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub metadata: Option<Metadata>,
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub struct Metadata {
+    pub name: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub timestamp: Option<String>,
+    #[serde(deserialize_with = "deserialize_string_from_number")]
+    #[serde(serialize_with = "maybe_serialize_string_to_integer")]
+    pub version: String,
```

Comment:
Seems like it might be better to have a `Version` struct instead of the manual serialization? Would also mean not needing the extra crate.

```
#[derive(Debug, serde::Serialize, serde::Deserialize)]
#[serde(untagged)]
pub enum Version {
    Number(i32),
    String(String),
}
```